File: attocode/src/shared/context-engine.ts
/**
 * Shared Context Engine (Phase 3.1)
 *
 * Orchestration layer that composes SharedContextState primitives into
 * worker-optimized prompts. Workers get a single buildWorkerSystemPrompt()
 * call instead of manually assembling prefix + failures + recitation.
 */

import type { SharedContextState } from './shared-context-state.js';
import type { FailureInput, Failure } from '../tricks/failure-evidence.js';
import type { Reference } from '../tricks/reversible-compaction.js';

// =============================================================================
// TYPES
// =============================================================================

export interface SharedContextEngineConfig {
  /** Maximum failures to include in worker prompts (default: 5) */
  maxFailuresInPrompt?: number;
  /** Whether to include cross-worker insights (default: true) */
  includeInsights?: boolean;
}

export interface WorkerTask {
  /** Unique task ID */
  id: string;
  /** Human-readable task description */
  description: string;
  /** The specific goal this worker should accomplish */
  goal: string;
  /** IDs of tasks this depends on */
  dependencies?: string[];
  /** Additional context for this specific task */
  context?: string;
}

// =============================================================================
// SHARED CONTEXT ENGINE
// =============================================================================

export class SharedContextEngine {
  private contextState: SharedContextState;
  private maxFailuresInPrompt: number;
  private includeInsights: boolean;

  constructor(contextState: SharedContextState, config: SharedContextEngineConfig = {}) {
    this.contextState = contextState;
    this.maxFailuresInPrompt = config.maxFailuresInPrompt ?? 5;
    this.includeInsights = config.includeInsights ?? true;
  }

  /**
   * Build a complete system prompt for a worker.
   * Concatenates: shared prefix → task context → failure guidance → goal recitation.
   * Empty sections are omitted.
   */
  buildWorkerSystemPrompt(task: WorkerTask): string {
    const sections: string[] = [];

    // 1. Shared prefix (cache-aligned, identical for all workers)
    const prefix = this.getSharedPrefix();
    if (prefix) {
      sections.push(prefix);
    }

    // 2. Task context
    const taskSection = this.buildTaskSection(task);
    if (taskSection) {
      sections.push(taskSection);
    }

    // 3. Failure guidance
    const failureGuidance = this.getFailureGuidance();
    if (failureGuidance) {
      sections.push(failureGuidance);
    }

    // 4. Goal recitation
    const goalRecitation = this.getGoalRecitation(task);
    if (goalRecitation) {
      sections.push(goalRecitation);
    }

    return sections.join('\n\n');
  }

  /**
   * Get the shared static prefix for KV-cache alignment.
   */
  getSharedPrefix(): string {
    return this.contextState.getStaticPrefix();
  }

  /**
   * Report a failure from a worker. Delegates to SharedContextState.
   */
  reportFailure(workerId: string, failure: FailureInput): Failure {
    return this.contextState.recordFailure(workerId, failure);
  }

  /**
   * Get formatted failure guidance for inclusion in prompts.
   * Returns empty string when no failures exist.
   */
  getFailureGuidance(): string {
    const parts: string[] = [];

    const failureContext = this.contextState.getFailureContext(this.maxFailuresInPrompt);
    if (failureContext) {
      parts.push(failureContext);
    }

    if (this.includeInsights) {
      const insights = this.contextState.getFailureInsights();
      if (insights.length > 0) {
        parts.push('## Cross-Worker Insights\n' + insights.map((i) => `- ${i}`).join('\n'));
      }
    }

    return parts.join('\n\n');
  }

  /**
   * Get a goal recitation block for a worker task.
   */
  getGoalRecitation(task: WorkerTask): string {
    const lines: string[] = [
      '## Current Goal',
      `**Task:** ${task.description}`,
      `**Goal:** ${task.goal}`,
    ];

    if (task.dependencies && task.dependencies.length > 0) {
      lines.push(`**Depends on:** ${task.dependencies.join(', ')}`);
    }

    return lines.join('\n');
  }

  /**
   * Search for relevant references. Delegates to SharedContextState.
   */
  getRelevantReferences(query: string): Reference[] {
    return this.contextState.searchReferences(query);
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  private buildTaskSection(task: WorkerTask): string {
    const parts: string[] = ['## Task Assignment'];

    parts.push(`You are working on task **${task.id}**: ${task.description}`);

    if (task.context) {
      parts.push(`\n### Additional Context\n${task.context}`);
    }

    return parts.join('\n');
  }
}

// =============================================================================
// FACTORY
// =============================================================================

export function createSharedContextEngine(
  contextState: SharedContextState,
  config?: SharedContextEngineConfig,
): SharedContextEngine {
  return new SharedContextEngine(contextState, config);
}

File: attocode/attocode_py/src/attocode/shared/context_engine.py
"""Shared context engine for building worker system prompts.

Composes SharedContextState into complete worker prompts by assembling
shared prefix, task context, failure guidance, and goal recitation.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from attocode.shared.shared_context_state import SharedContextState


@dataclass(slots=True)
class SharedContextEngineConfig:
    """Configuration for shared context engine."""

    max_failure_context: int = 10
    include_insights: bool = True
    include_references: bool = True


@dataclass(slots=True)
class WorkerTask:
    """Task specification for a worker."""

    id: str
    description: str
    goal: str
    dependencies: list[str] = field(default_factory=list)
    context: str = ""


@dataclass
class SharedContextEngine:
    """Orchestration layer composing shared state into worker prompts.

    Builds complete worker system prompts by assembling:
    - Shared prefix for KV-cache alignment
    - Task-specific context
    - Failure guidance from cross-worker learning
    - Goal recitation for persistence
    """

    shared_state: SharedContextState = field(default_factory=SharedContextState)
    config: SharedContextEngineConfig = field(default_factory=SharedContextEngineConfig)

    def build_worker_system_prompt(self, task: WorkerTask) -> str:
        """Create full system prompt for a worker."""
        parts: list[str] = []

        # Shared prefix for cache alignment
        prefix = self.get_shared_prefix()
        if prefix:
            parts.append(prefix)

        # Task context
        parts.append(self._build_task_context(task))

        # Failure guidance
        guidance = self.get_failure_guidance()
        if guidance:
            parts.append(guidance)

        # Goal recitation
        parts.append(self.get_goal_recitation(task))

        return "\n\n".join(parts)

    def get_shared_prefix(self) -> str:
        """Static prefix for cache alignment."""
        prefix = self.shared_state.kv_cache_prefix
        if prefix:
            return f"<shared-context>\n{prefix}\n</shared-context>"
        return ""

    def report_failure(self, worker_id: str, failure: str, error: str = "") -> None:
        """Delegate to SharedContextState."""
        self.shared_state.record_failure(worker_id, failure, error)

    def get_failure_guidance(self) -> str:
        """Format failure context and cross-worker insights."""
        context = self.shared_state.get_failure_context(
            max_failures=self.config.max_failure_context
        )

        if not context:
            return ""

        parts = ["<failure-guidance>", context]

        if self.config.include_insights:
            insights = self.shared_state.get_failure_insights()
            if insights:
                parts.append("\n**Key Insights:**")
                for insight in insights:
                    parts.append(f"- {insight}")

        parts.append("</failure-guidance>")
        return "\n".join(parts)

    def get_goal_recitation(self, task: WorkerTask) -> str:
        """Build goal block for a worker."""
        parts = [
            "<goal>",
            f"Task: {task.description}",
            f"Goal: {task.goal}",
        ]

        if task.dependencies:
            parts.append(f"Dependencies: {', '.join(task.dependencies)}")

        if task.context:
            parts.append(f"Context: {task.context}")

        parts.append("</goal>")
        return "\n".join(parts)

    def get_relevant_references(self, query: str) -> list[Any]:
        """Search pooled references."""
        if not self.config.include_references:
            return []
        return self.shared_state.search_references(query)

    def _build_task_context(self, task: WorkerTask) -> str:
        """Build task-specific context section."""
        return (
            f"<task id=\"{task.id}\">\n"
            f"Description: {task.description}\n"
            f"Goal: {task.goal}\n"
            "</task>"
        )
